electrolytes/solvation_shell_extract.py

```diff
@@ -0,0 +1,178 @@
+
+    logging.info("Translating atoms to solute center of mass")
+    for ts in tqdm(universe.trajectory):
+        ts.dimensions = universe.dimensions
+        solu_center = solu.center_of_mass(wrap=True)
+        dim = ts.triclinic_dimensions
+        box_center = np.sum(dim, axis=0) / 2
+        universe.atoms.translate(box_center - solu_center)
+
+        universe.atoms.unwrap()
+
+    solvent = universe.atoms - solu
+
+    solv_anal = Solute.from_atoms(solu, {"solvent": solvent}, solute_name=solute_atom)
+
+    # Identify the cutoff for the first solvation shell, based on the MD trajectory
+    logging.info("Running solvation analysis")
+    solv_anal.run()
+
+    # Plot the RDF
+    solv_anal.plot_solvation_radius("solute", "solvent")
+    plt.savefig(os.path.join(save_dir, system_name, solute_atom, "solvation_rdf.png"))
+
+    # There's probably a much faster way to do this
+    # But for now, we're prototyping, so slow is okay
+    shells = dict()
+    for j in solv_anal.speciation.speciation_fraction["solvent"]:
+        shells[j] = solv_anal.speciation.get_shells({"solvent": j})
+
+    # Now let's try getting the most diverse structures for each particular coordination number
+    # This is also a bit slow, particularly for the more common and/or larger solvent shells
+    for c in range(min_coord, max_coord + 1):
+        logging.info(f"Processing shells with coordination number {c}")
+        os.makedirs(
+            os.path.join(save_dir, system_name, solute_atom, f"coord={c}"),
+            exist_ok=True,
+        )
+        shell_species = []
+        shell_positions = []
+        for index, _ in tqdm(shells[c].iterrows()):
+            ts = universe.trajectory[index[0]]
+            universe.atoms.unwrap()
+            shell = solv_anal.solvation_data.xs(
+                (index[0], index[1]), level=(FRAME, SOLUTE_IX)
+            )
+            shell = solv_anal._df_to_atom_group(shell, solute_index=index[1])
+            shell = shell.copy()
+            if len(shell.atoms.elements) > len(shell_species):
+                shell_species = shell.atoms.elements
+
+            shell_positions.append(wrap_positions(shell.atoms.positions, lattices))
+
+        by_num_atoms = defaultdict(list)
+        for sps in shell_positions:
+            by_num_atoms[len(sps)].append(sps)
+
+        # filter by number of atoms per shell
+        selections_by_num_atoms = {
+            num_atoms: filter_by_rmse(shells_with_num_atoms, top_n)
+            for num_atoms, shells_with_num_atoms in by_num_atoms.items()
+        }
+
+        for (
+            shell_size,
+            shell_positions,
+        ) in selections_by_num_atoms.items():  # loop over sizes
+            for idx, shell_pos in enumerate(shell_positions):
+                if shell_pos.shape[0] == shell_species.shape[0]:
+
+                    # Save shell as xyz file
+                    mol = Molecule(shell_species, shell_pos, charge=-1)
```

Comment:
To further emphasize, you should be checking partial charges/charges of the various residues to assign charge here.

Reply:
Hm, is there an easy way to get charge info from the PDB file? Or from MD Analysis?

Reply:
Uhhh PDB CAN include charges. Likewise, MDAnalysis Universe objects include partial charges, if they're present in the trajectory. I assume that OpenMM stores that information? We should check on this.

Reply:
I don't think the pdb file which Muhammad shared with me (see `testfiles/water_nacl_example.pdb`) has this information. I can follow up with him on this.